### PR TITLE
Mark parametrize get_example_tree_sequences in tests

### PR DESCRIPTION
**Pull request by hyanwong**
Seems sensible to move to new method. We can do caching etc later.

### CODE REVIEW

**Review by jeromekelleher — APPROVED** [review]
Very nice :+1:

**Comment by hyanwong** [comment]
Something wacky going on in Windows here with memory usage in the fixtures. Perhaps it could be solved by marking the tree sequences generated by `get_example_tree_sequences` as fixed? Any suggestions e.g. @benjeffery ?

```
Run source ~/.profile
[7](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:8)
============================= test session starts =============================
[8](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:9)
platform win32 -- Python 3.10.4, pytest-7.1.1, pluggy-1.0.0
[9](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:10)
rootdir: D:\a\tskit\tskit\python, configfile: pytest.ini, testpaths: tests
[10](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:11)
plugins: cov-3.0.0, forked-1.4.0, xdist-2.5.0
[11](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:12)
gw0 I / gw1 I
[12](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:13)

[13](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:14)
=================================== ERRORS ====================================
[14](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:15)
________________ ERROR collecting tests/test_phylo_formats.py _________________
[15](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:16)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\pluggy\_hooks.py:265: in __call__
[16](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:17)
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
[17](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:18)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\pluggy\_manager.py:80: in _hookexec
[18](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:19)
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
[19](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:20)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\_pytest\python.py:262: in pytest_pycollect_makeitem
[20](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:21)
    return list(collector._genfunctions(name, obj))
[21](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:22)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\_pytest\python.py:489: in _genfunctions
[22](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:23)
    self.ihook.pytest_generate_tests.call_extra(methods, dict(metafunc=metafunc))
[23](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:24)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\pluggy\_hooks.py:292: in call_extra
[24](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:25)
    return self(**kwargs)
[25](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:26)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\pluggy\_hooks.py:265: in __call__
[26](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:27)
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
[27](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:28)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\pluggy\_manager.py:80: in _hookexec
[28](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:29)
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
[29](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:30)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\_pytest\python.py:148: in pytest_generate_tests
[30](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:31)
    metafunc.parametrize(*marker.args, **marker.kwargs, _param_mark=marker)
[31](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:32)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\_pytest\python.py:1272: in parametrize
[32](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:33)
    argnames, parametersets = ParameterSet._for_parametrize(
[33](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:34)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\_pytest\mark\structures.py:160: in _for_parametrize
[34](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:35)
    parameters = cls._parse_parametrize_parameters(argvalues, force_tuple)
[35](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:36)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\_pytest\mark\structures.py:146: in _parse_parametrize_parameters
[36](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:37)
    return [
[37](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:38)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\_pytest\mark\structures.py:146: in <listcomp>
[38](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:39)
    return [
[39](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:40)
tests\test_highlevel.py:289: in get_example_tree_sequences
[40](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:41)
    ts = msprime.simulate(
[41](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:42)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\msprime\ancestry.py:609: in simulate
[42](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:43)
    sim = _parse_simulate(
[43](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:44)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\msprime\ancestry.py:358: in _parse_simulate
[44](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:45)
    sim = Simulator(
[45](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:46)
C:\Miniconda\envs\anaconda-client-env\lib\site-packages\msprime\ancestry.py:1318: in __init__
[46](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:47)
    super().__init__(
[47](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:48)
E   msprime._msprime.InputError: Input error in initialise: Out of memory.
[48](https://github.com/tskit-dev/tskit/runs/7993242594?check_suite_focus=true#step:13:49)
```

**Comment by benjeffery** [comment]
I've amended the commit to add caching to `get_example_tree_sequences`, lets see how windows likes that.

**Comment by benjeffery** [comment]
@mergifyio rebase

**Review by benjeffery — APPROVED** [review]
Nice

**Comment by benjeffery** [comment]
Hmm reusing the tree sequences is making some tests fail. The problem here is that pytest is calling `get_example_tree_sequences` as part of it's init and storing them all for the duration of the run, hence running out of memory.
I think the fix is to get `get_example_tree_sequences` to return copies.

**Comment by petrelharp** [comment]
It looks like the iterator isn't iterating? Is there something weird about cacheing an iterator? (good thing we had that test that checks if any tests had actually been run - maybe we also need to add a test to test that this cached fixture is actually working? I'm not sure how to do this, though)

**Comment by benjeffery** [comment]
Ok, I've fixed this so that we only have one copy of the example tree sequences. This reduces test time by about 20%!!!!

File: python/tests/test_highlevel.py [hunk 1]
```diff
@@ -225,20 +228,23 @@ def get_gap_examples():
...
     tables = get_table_collection_copy(ts.dump_tables(), 2)
     tables.sites.clear()
     tables.mutations.clear()
     insert_uniform_mutations(tables, 100, list(ts.samples()))
-    yield tables.tree_sequence()
+    ret.append(("gap at end", tables.tree_sequence()))
+    return ret
 
 
+@functools.lru_cache()
```
Comment (petrelharp) [inline python/tests/test_highlevel.py:242]: Did you mean to cache this one and have it return a list, while `get_example_tree_sequences` doesn't have a cache decorator and most of the other build-ts-functions are still iterators?

Comment (benjeffery) [reply python/tests/test_highlevel.py:242]: I converted the ones called from other places other than `get_example_tree_sequences`. `get_example_tree_sequences` is cached, but manually.

**Pull request merged**